File: packages/ts/cli/src/schema-bridge.ts
/**
 * Minimal JSON-Schema → Zod converter for the `structured` command. The
 * orchestrator's `generateStructured` enforces output against a Zod schema, but
 * the CLI accepts a JSON Schema file from `--schema`; this is the bridge. It is
 * the inverse of `@llmorch/core`'s in-tree `zodToJsonSchema` and covers the same
 * subset (object/string/number/integer/boolean/array/enum + `required`) — kept
 * dependency-free in line with the repo's zero-extra-runtime-deps policy.
 */
import { z, type ZodTypeAny } from "zod";

interface JsonSchemaNode {
  type?: string;
  properties?: Record<string, JsonSchemaNode>;
  required?: string[];
  items?: JsonSchemaNode;
  enum?: unknown[];
}

export function jsonSchemaToZod(schema: JsonSchemaNode): ZodTypeAny {
  if (Array.isArray(schema.enum) && schema.enum.every((v) => typeof v === "string")) {
    return z.enum(schema.enum as [string, ...string[]]);
  }
  switch (schema.type) {
    case "object": {
      const required = new Set(schema.required ?? []);
      const shape: Record<string, ZodTypeAny> = {};
      for (const [key, child] of Object.entries(schema.properties ?? {})) {
        const inner = jsonSchemaToZod(child);
        shape[key] = required.has(key) ? inner : inner.optional();
      }
      return z.object(shape);
    }
    case "string":
      return z.string();
    case "integer":
      return z.number().int();
    case "number":
      return z.number();
    case "boolean":
      return z.boolean();
    case "array":
      return z.array(schema.items ? jsonSchemaToZod(schema.items) : z.unknown());
    default:
      return z.unknown();
  }
}

File: packages/py/llmorch-cli/src/llmorch_cli/schema_bridge.py
"""Minimal JSON-Schema → Pydantic model builder — twin of schema-bridge.ts.

The orchestrator's ``generate_structured`` validates output against a Pydantic
model class, but the CLI accepts a JSON Schema file from ``--schema``; this is the
bridge. It covers the same subset as ``llmorch_core``'s ``model_to_json_schema``
(object/string/number/integer/boolean/array + ``required``) and is dependency-free
in line with the repo's zero-extra-runtime-deps policy.
"""

from typing import Any

from pydantic import BaseModel, create_model

_JSON_TO_PY: dict[str, type] = {
    "string": str,
    "integer": int,
    "number": float,
    "boolean": bool,
}


def _type_for(node: dict[str, Any]) -> Any:
    kind = node.get("type")
    if kind == "object":
        return json_schema_to_model(node)
    if kind == "array":
        items = node.get("items")
        return list[_type_for(items)] if isinstance(items, dict) else list  # type: ignore[misc]
    return _JSON_TO_PY.get(kind, Any) if isinstance(kind, str) else Any


def json_schema_to_model(schema: dict[str, Any], name: str = "StructuredOutput") -> type[BaseModel]:
    """Build a Pydantic model class from a (subset of) JSON Schema object."""
    properties: dict[str, Any] = schema.get("properties", {})
    required = set(schema.get("required", []))
    fields: dict[str, Any] = {}
    for key, sub in properties.items():
        py_type = _type_for(sub)
        if key in required:
            fields[key] = (py_type, ...)
        else:
            fields[key] = (py_type | None, None)
    return create_model(name, **fields)
